libflux/src/semantic/types.rs

```diff
@@ -842,8 +842,71 @@ impl Substitutable for Function {
 }
 
 impl Function {
-    fn unify(self, _: Self, _: &mut TvarKinds) -> Result<Substitution, Error> {
-        unimplemented!();
+    // In order to unify two function types, it is necessary and sufficient to unify their argument and return types.
+    // This seems straight forward however there are some subtleties when unifying argument types.
+    fn unify(
+        self,
+        other: Self,
+        cons: &mut TvarKinds,
+        fresh: &mut Fresher,
+    ) -> Result<Substitution, Error> {
+        let f = self;
+        let mut g = other;
+        // We start with a validation phase from steps 1 and 3:
+        //  1. Verify that a required argument of f is in the required/optional arguments of g (and viceversa).
+        //  3. If a pipe arg was used then the other type must specify a pipe arg too.
+        // This is done at the beginning both to avoid starting unifying when there is no need to,
+        // and in order to error using f and g, otherwise, the borrow checker would not make us
+        // error while we have borrowed some of their internal pieces (while iterating for example).
+        if f.req
+            .keys()
+            .any(|k| !g.req.contains_key(k) && !g.opt.contains_key(k))
+            || g.req
+                .keys()
+                .any(|k| !f.req.contains_key(k) && !f.opt.contains_key(k))
+            || match (&f.pipe, &g.pipe) {
+                (Some(_), None) | (None, Some(_)) => true,
+                _ => false,
+            }
+        {
+            return Err(Error::cannot_unify(&f, &g));
+        }
+        // Given two function types f and g, the process for unifying their arguments is as follows:
+        // 1. If a required arg of f is not present in the arguments of g, fail unification (checked above),
+        //    otherwise unify both argument types.
+        let sub: Substitution =
+            f.req
+                .into_iter()
+                .try_fold(Substitution::empty(), |sub, (arg_name, f_arg_type)| {
+                    if let Some(g_arg_type) = g.req.remove(arg_name.as_str()) {
```

Comment:
I am using `remove` because that's the only way to own something from an HashMap